### PR TITLE
Add Keyring Feature + Embed Chapters + Fix Validation

### LINKED ISSUE #327: Implement Keyring for Usernames and Passwords
Use [Nickvision.Keyring](https://github.com/NickvisionApps/Keyring)

### LINKED ISSUE #338: Enable YouTube Chapter Support
Another app that fulfills the same propose as TubeConverter, VideoDownloader, includes chapters (or timestamps) of downloaded videos. 

Both use yt-dlp to power the program, so TubeConverter can support it. It's especially useful for educational videos that use chapters/timestamps to divide the content. 

I would include a toggle called "Download chapters for supported media".

### LINKED ISSUE #340: "Media URL (Invalid)" error on some URLs
Using the Flathub version on Fedora Silverblue 38

Although YouTube addresses work fine, URLs from [ina.fr](https://www.ina.fr/) seem to have stopped being recognized as valid. Feeding those to ytp-dl from the command line does work though.

```
$ flatpak --env=LANG=en_US run org.nickvision.tubeconverter
Python.Runtime.PythonException: 'formats'
   at Python.Runtime.PythonException.ThrowLastAsClrException()
   at Python.Runtime.PyObject.GetItem(PyObject key)
   at Python.Runtime.PyObject.GetItem(String key)
   at Python.Runtime.PyObject.get_Item(String key)
   at NickvisionTubeConverter.Shared.Models.MediaUrlInfo.ParseFromPyDict(PyDict mediaInfo, Boolean isPartOfPlaylist) in /run/build/org.nickvision.tubeconverter/NickvisionTubeConverter.Shared/Models/MediaUrlInfo.cs:line 112
   at NickvisionTubeConverter.Shared.Models.MediaUrlInfo.<>c__DisplayClass13_0.<GetAsync>b__0() in /run/build/org.nickvision.tubeconverter/NickvisionTubeConverter.Shared/Models/MediaUrlInfo.cs:line 82
```

image

### PR DESCRIPTION
**Pull request by nlogozzo**
Closes #327 
Closes #338 
Closes #340

### CODE REVIEW

**Comment by nlogozzo** [comment]
Ok, KeyringDialog is all done. You should be able to unlock Keyring on startup if enabled and then in the dialog, add, edit, and delete credentials.
I have to set out now so then later tonight when I get back I will add selecting credentials in AddDownloadDialog.
The `Enable Chapters` preference is also complete and in this PR f you want to test that as well :)

**Comment by DaPigGuy** [comment]
Could we look into using the [secrets portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Secret) for the keyring password?

**Comment by nlogozzo** [comment]
> Could we look into using the [secrets portal](https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.Secret) for the keyring password?

For now no. We'd like to keep Nickvision.Keyring portable across all platforms, making the keyrings indepdent of platforms secrets and compatible across

**Comment by nlogozzo** [comment]
AddDownloadDialog with Keyring:
image

Manual support still supported:
image

If keyring is not enabled, you can still manually authenticate:
image

(Thanks @soumyaDghosh for the suggestion to turn authentication into ExpanderRow)

**Comment by nlogozzo** [comment]
PR ready :)

**Comment by nlogozzo** [comment]
Actually, I would like to add an enabled switch, that way a credential doesn't get passed if not needed.

**Comment by nlogozzo** [comment]
image

OK, now ready :)

File: NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs [hunk 1]
```diff
@@ -296,7 +321,14 @@ private AddDownloadDialog(Gtk.Builder builder, AddDownloadDialogController contr
                 quality = Quality.Resolution;
                 resolution = _mediaUrlInfo.VideoResolutions[(int)_qualityRow.GetSelected()];
             }
-            _controller.PopulateDownloads(_mediaUrlInfo!, fileType, quality, resolution, (Subtitle)_subtitleRow.GetSelected(), _saveFolderString, _speedLimitSwitch.GetActive(), _cropThumbnailSwitch.GetActive(), _usernameRow.GetText(), _passwordRow.GetText());
+            if(_keyringRow.GetSelected() == 0 || _keyringRow.GetSelected() == 4294967295 || !_authRow.GetEnableExpansion())
```
Comment (fsobolev) [inline NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs:324]: Maybe `uint.MaxValue` or at least a comment would be better than just magic number?

File: NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs [hunk 2]
```diff
@@ -374,7 +423,14 @@ private async Task SearchUrlAsync(string url)
             try
             {
                 _urlRow.SetText(url);
-                _mediaUrlInfo = await _controller.SearchUrlAsync(url, _usernameRow.GetText(), _passwordRow.GetText());
+                if(_keyringRow.GetSelected() == 0 || _keyringRow.GetSelected() == 4294967295 || !_authRow.GetEnableExpansion())
```
Comment (fsobolev) [inline NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs:426]: Same here

**Review by fsobolev — CHANGES_REQUESTED** [review]

File: NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs (hunk 1 above)
Comment (nlogozzo) [reply NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs:324]: The magic number is just https://docs.gtk.org/gtk4/const.INVALID_LIST_POSITION.html

Probably best to just add a comment, could you do it?

Comment (nlogozzo) [reply NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs]: Ok, fixed with https://github.com/NickvisionApps/TubeConverter/pull/330/commits/9acc846e896db454635d0876c395f9a56c2f45b1

File: NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs (hunk 2 above)
Comment (nlogozzo) [reply NickvisionTubeConverter.GNOME/Views/AddDownloadDialog.cs]: https://github.com/NickvisionApps/TubeConverter/pull/330/commits/9acc846e896db454635d0876c395f9a56c2f45b1

**Review by fsobolev — APPROVED** [review]

**Review by DaPigGuy — APPROVED** [review]

**Pull request merged**